replication/event.go

```diff
@@ -229,15 +229,59 @@ type PreviousGTIDsEvent struct {
 	GTIDSets string
 }
 
+type GtidFormat int
+
+const (
+	GtidFormatClassic = iota
+	GtidFormatTagged
+)
+
+func decodeSid(data []byte) (format GtidFormat, sidnr uint64) {
+	if data[7] == 1 {
+		format = GtidFormatTagged
+	}
+
+	if format == GtidFormatTagged {
+		sid_mask := []byte{0, 255, 255, 255, 255, 255, 255, 0}
+
+		// Apply the mask
+		for i, _ := range data[:8] {
+			data[i] &= sid_mask[i]
+		}
+		data = append(data, 0)
+
+		// sidnr
+		sidnr = binary.LittleEndian.Uint64(data[1:])
+		return
+	}
+	sidnr = binary.LittleEndian.Uint64(data)
+	return
+}
+
 func (e *PreviousGTIDsEvent) Decode(data []byte) error {
 	pos := 0
-	uuidCount := binary.LittleEndian.Uint16(data[pos : pos+8])
+
+	gtidinfo := make([]byte, 8)
+	copy(gtidinfo, data[:8])
+	format, uuidCount := decodeSid(gtidinfo)
 	pos += 8
 
 	previousGTIDSets := make([]string, uuidCount)
-	for i := range previousGTIDSets {
+	currentSetnr := 0
+	for _ = range previousGTIDSets {
 		uuid := e.decodeUuid(data[pos : pos+16])
 		pos += 16
+		isTag := false
+		var tag string
+		if format == GtidFormatTagged {
+			tagLength := int(data[pos]) / 2
+			pos += 1
+			if tagLength > 0 {
```

Comment:
can the tag have zero length? please add comments

Reply:
```
mysql-9.1.0> set gtid_next = 'AUTOMATIC:abc';
Query OK, 0 rows affected (0.00 sec)

mysql-9.1.0> create table test.t9 (id int primary key);
Query OK, 0 rows affected (0.02 sec)

mysql-9.1.0> set gtid_next = 'AUTOMATIC:';
ERROR 1774 (HY000): Malformed GTID specification 'AUTOMATIC:'.
```

https://dev.mysql.com/doc/dev/mysql-server/latest/classmysql_1_1gtid_1_1Tag.html#details

> Representation of the GTID tag.
>
> Tag format: [a-z_]{a-z0-9_}{0,31} Tag may be created from a given text. Text may contain leading and trailing spaces which will be omitted during Tag creation. Text may also contain uppercase characters. Acceptable format for text is as follows: [:space:][a-zA-Z_][a-zA-Z0-9_]{0,31}[:space:]

So it is non-zero if it has a tag. Note the `{0,31}` in the regexp, which is only for the second character as the first can't be a number.